File: Provenance/Controller/PVButtonGroupOverlayView.m
//
//  PVButtonOverlayView.m
//  Provenance
//
//  Created by James Addyman on 17/08/2013.
//  Copyright (c) 2013 James Addyman. All rights reserved.
//

#import "PVButtonGroupOverlayView.h"
#import "JSButton.h"

@interface PVButtonGroupOverlayView ()

@property (nonatomic, strong) NSArray *buttons;

@end

@implementation PVButtonGroupOverlayView

- (id)initWithButtons:(NSArray *)buttons
{
	if ((self = [super initWithFrame:CGRectZero]))
	{
#if !TARGET_OS_TV
		[self setMultipleTouchEnabled:YES];
#endif
		[self setBackgroundColor:[UIColor clearColor]];
		self.buttons = buttons;
	}
	
	return self;
}

- (void)dealloc
{
	self.buttons = nil;
}

- (void)willMoveToSuperview:(UIView *)newSuperview
{
	[super willMoveToSuperview:newSuperview];
}

- (void)touchesBegan:(NSSet *)touches withEvent:(UIEvent *)event
{
	UITouch *touch = [touches anyObject];
	CGPoint location = [touch locationInView:self];
	
	for (JSButton *button in [self buttons])
	{
        CGRect touchArea = CGRectMake(location.x - 10, location.y - 10, 20, 20);
        CGRect buttonFrame = [self convertRect:[button frame] toView:self];
        if (CGRectIntersectsRect(touchArea, buttonFrame))
//		if (CGRectContainsPoint([button frame], location))
		{
			[button touchesBegan:touches withEvent:event];
		}
	}
}

- (void)touchesMoved:(NSSet *)touches withEvent:(UIEvent *)event
{
	UITouch *touch = [touches anyObject];
	CGPoint location = [touch locationInView:self];
	
	for (JSButton *button in [self buttons])
	{
        CGRect touchArea = CGRectMake(location.x - 10, location.y - 10, 20, 20);
        CGRect buttonFrame = [self convertRect:[button frame] toView:self];
        if (CGRectIntersectsRect(touchArea, buttonFrame))
//		if (CGRectContainsPoint([button frame], location))
		{
			[button touchesMoved:touches withEvent:event];
		}
		else if (button.pressed)
		{
			[button touchesEnded:touches withEvent:event];
		}
	}
}

- (void)touchesCancelled:(NSSet *)touches withEvent:(UIEvent *)event
{
	for (JSButton *button in [self buttons])
	{
		[button touchesCancelled:touches withEvent:event];
	}
}

- (void)touchesEnded:(NSSet *)touches withEvent:(UIEvent *)event
{
	for (JSButton *button in [self buttons])
	{
		[button touchesEnded:touches withEvent:event];
	}
}

@end

File: Provenance/Controller/PVButtonGroupOverlayView.swift
//
//  PVButtonOverlayView.swift
//  Provenance
//
//  Created by Joe Mattiello on 17/03/2018.
//  Copyright (c) 2018 Joe Mattiello. All rights reserved.
//

#if canImport(UIKit)
import UIKit
#else
import AppKit
#endif
import PVLogging

final class PVButtonGroupOverlayView: MovableButtonView {
    var buttons = [JSButton]()

    init(buttons: [JSButton]) {
        super.init(frame: CGRect.zero)

#if os(iOS)
        isMultipleTouchEnabled = true
#endif
        backgroundColor = UIColor.clear
        self.buttons = buttons
    }

    required init?(coder _: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        if inMoveMode {
            super.touchesBegan(touches, with: event)
            return
        }
        let touch = touches.first!
        let location = touch.location(in: self)
        for button: JSButton in buttons {
            let touchArea = CGRect(x: location.x - 10, y: location.y - 10, width: 20, height: 20)
            let buttonFrame: CGRect = convert(button.frame, to: self)
            if touchArea.intersects(buttonFrame) {
                button.touchesBegan(touches, with: event)
            }
        }
    }

    override func touchesMoved(_ touches: Set<UITouch>, with event: UIEvent?) {
        if inMoveMode {
            super.touchesMoved(touches, with: event)
            return
        }
		guard let touch = touches.first else {
			ELOG("No touches!?")
			return
		}
        let location = touch.location(in: self)
        for button: JSButton in buttons {
            let touchArea = CGRect(x: location.x - 10, y: location.y - 10, width: 20, height: 20)
            let buttonFrame: CGRect = convert(button.frame, to: self)
            if touchArea.intersects(buttonFrame) {
                button.touchesMoved(touches, with: event)
            } else if button.pressed {
                button.touchesEnded(touches, with: event)
            }
        }
    }

    override func touchesCancelled(_ touches: Set<UITouch>, with event: UIEvent?) {
        if inMoveMode {
            super.touchesCancelled(touches, with: event)
            return
        }
        for button: JSButton in buttons {
            button.touchesCancelled(touches, with: event)
        }
    }

    override func touchesEnded(_ touches: Set<UITouch>, with event: UIEvent?) {
        if inMoveMode {
            super.touchesEnded(touches, with: event)
            return
        }
        for button: JSButton in buttons {
            button.touchesEnded(touches, with: event)
        }
    }

    var isMovable: Bool {
        return true
    }

	override func didStartMoving() {
		super.didStartMoving()
		buttons.forEach { $0.isUserInteractionEnabled = false }
	}

	override func didFinishMoving(velocity:CGPoint) {
		super.didFinishMoving(velocity: velocity)
		buttons.forEach { $0.isUserInteractionEnabled = true }
	}
}
